pkg/dataplane/template.go

```diff
@@ -0,0 +1,214 @@
+		if maxOpenFiles < 1024 {
+			// this means the value of RLIMIT_NOFILE is too low.
+			maxOpenFiles = 1024
+		}
+		klog.V(3).InfoS("Adjusting MaxWorkerOpenFiles variable", "value", maxOpenFiles)
+		cfg.MaxWorkerOpenFiles = maxOpenFiles
+	}
+
+	if cfg.MaxWorkerConnections == 0 {
+		maxWorkerConnections := int(float64(cfg.MaxWorkerOpenFiles * 3.0 / 4))
+		klog.V(3).InfoS("Adjusting MaxWorkerConnections variable", "value", maxWorkerConnections)
+		cfg.MaxWorkerConnections = maxWorkerConnections
+	}
+
+	setHeaders := map[string]string{}
+	if cfg.ProxySetHeaders != "" {
+		// TODO: Get via gRPC
+		// cmap, err := n.store.GetConfigMap(cfg.ProxySetHeaders)
+		cmap, err := &v1.ConfigMap{}, fmt.Errorf("")
+		if err != nil {
+			klog.Warningf("Error reading ConfigMap %q from local store: %v", cfg.ProxySetHeaders, err)
+		} else {
+			setHeaders = cmap.Data
+		}
+	}
+
+	addHeaders := map[string]string{}
+	if cfg.AddHeaders != "" {
+		// TODO: Get via gRPC
+		// cmap, err := n.store.GetConfigMap(cfg.AddHeaders)
+		cmap, err := &v1.ConfigMap{}, fmt.Errorf("")
+		if err != nil {
+			klog.Warningf("Error reading ConfigMap %q from local store: %v", cfg.AddHeaders, err)
+		} else {
+			addHeaders = cmap.Data
+		}
+	}
+
+	sslDHParam := ""
+	if cfg.SSLDHParam != "" {
+		secretName := cfg.SSLDHParam
+
+		// TODO: get via gRPC or build a cache
+		//secret, err := n.store.GetSecret(secretName)
+		secret, err := &v1.Secret{}, fmt.Errorf("")
+		if err != nil {
+			klog.Warningf("Error reading Secret %q from local store: %v", secretName, err)
+		} else {
+			nsSecName := strings.Replace(secretName, "/", "-", -1)
+			dh, ok := secret.Data["dhparam.pem"]
+			if ok {
+				pemFileName, err := ssl.AddOrUpdateDHParam(nsSecName, dh)
+				if err != nil {
+					klog.Warningf("Error adding or updating dhparam file %v: %v", nsSecName, err)
+				} else {
+					sslDHParam = pemFileName
+				}
+			}
+		}
+	}
+
+	cfg.SSLDHParam = sslDHParam
+
+	cfg.DefaultSSLCertificate = n.getDefaultSSLCertificate()
+
+	tc := ngx_config.TemplateConfig{
+		ProxySetHeaders:          setHeaders,
+		AddHeaders:               addHeaders,
+		BacklogSize:              sysctlSomaxconn(),
+		Backends:                 ingressCfg.Backends,
+		PassthroughBackends:      ingressCfg.PassthroughBackends,
+		Servers:                  ingressCfg.Servers,
+		TCPBackends:              ingressCfg.TCPEndpoints,
+		UDPBackends:              ingressCfg.UDPEndpoints,
+		Cfg:                      cfg,
+		IsIPV6Enabled:            n.isIPV6Enabled && !cfg.DisableIpv6,
+		NginxStatusIpv4Whitelist: cfg.NginxStatusIpv4Whitelist,
+		NginxStatusIpv6Whitelist: cfg.NginxStatusIpv6Whitelist,
+		RedirectServers:          utilingress.BuildRedirects(ingressCfg.Servers),
+		IsSSLPassthroughEnabled:  n.cfg.EnableSSLPassthrough,
+		ListenPorts:              n.cfg.ListenPorts,
+		// TODO: Is this used in the template? I don't think so
+		//PublishService:           n.GetPublishService(),
```

Comment:
IIRC, this is not used in the template. It can be used to identify the access address.

For our actual situation, we may use the following two places:

* Update the status of the Ingress resource
* Update the state of the Gateway resource in the Gateway API

Reply:
Cleaned in another PR :)